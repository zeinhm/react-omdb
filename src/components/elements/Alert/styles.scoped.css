.alert-component {
  align-self: center;
  display: flex;
  flex-direction: row;
  justify-content: center;
}

@media only screen and (max-width: 48em) {
  .alert-component {
    margin: 0 1rem;
  }
}