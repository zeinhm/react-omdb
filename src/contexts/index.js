import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AppContext = React.createContext(null);
export const AlertContext = React.createContext(null);
export const ModalContext = React.createContext(null);

export default function AppContextProvider({ children }) {
  const [alert, setAlert] = useState(null);
  const [breadcrumbs, setBreadcrumbs] = useState([]);
  const [content, setContent] = useState(null);
  const appValue = {
    breadcrumbs,
    setBreadcrumbs,
  };
  const alertValue = {
    alert,
    setAlert,
  };
  const modalValue = {
    content,
    setContent,
  };

  return (
    <AppContext.Provider value={appValue}>
      <AlertContext.Provider value={alertValue}>
        <ModalContext.Provider value={modalValue}>
          {children}
        </ModalContext.Provider>
      </AlertContext.Provider>
    </AppContext.Provider>
  );
}

AppContextProvider.defaultProps = {
  children: null,
};

AppContextProvider.propTypes = {
  children: PropTypes.node,
};
