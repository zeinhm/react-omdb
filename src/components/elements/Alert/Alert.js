import React, { useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { AlertContext } from '../../../contexts';
import './styles.scoped.css';

export default function Alert(props) {
  const { alert, setAlert } = useContext(AlertContext);
  const { open, className, children, onClose } = props;

  useEffect(() => {
    if (open) setAlert(<Content children={children} className={className} />);
    else setAlert(null);
  }, [open]);

  const [alertContent, setAlertContent] = useState(alert);

  useEffect(() => {
    if (alertContent && !alert) {
      onClose();
      setAlertContent(alert);
    } else if (!alertContent && alert) {
      setAlertContent(alert);
    }
  });

  return null;
}

Alert.defaultProps = {
  children: null,
  className: '',
  onClose: () => { },
  open: false,
};

alert.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  onClose: PropTypes.func,
  open: PropTypes.bool,
};

export function Content({ className, children }) {
  const customClass = ['alert-component', className].filter(Boolean).join(' ');

  return (
    <section className={customClass}>
      {children}
    </section>
  );
}

Content.defaultProps = {
  children: null,
  className: '',
};

Content.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
};
