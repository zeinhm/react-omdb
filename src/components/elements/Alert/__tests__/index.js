import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Alert, { Content } from '../Alert';
import { AlertContext } from '../../../../contexts';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: v => v,
  useEffect: cb => cb(),
  useState: v => [v === 'tes' ? v : !v, jest.fn()],
}));

jest.mock('../../../../contexts', () => ({
  AlertContext: {
    alert: '',
    setAlert: jest.fn(),
  },
}));

describe('src/components/elements/Alert', () => {
  test('render', () => {
    Alert.defaultProps.onClose();
    Alert.defaultProps.onClose = jest.fn();
    Alert(Alert.defaultProps);
    expect(AlertContext.setAlert).toHaveBeenNthCalledWith(1, null);
    expect(Alert.defaultProps.onClose).toHaveBeenCalled();

    const openProps = { ...Alert.defaultProps, open: true };
    const { className, children } = Alert.defaultProps;
    const contentProps = { children, className, };
    AlertContext.alert = 'test';
    Alert(openProps);
    expect(AlertContext.setAlert).toHaveBeenNthCalledWith(2, <Content {...contentProps} />);

    AlertContext.alert = 'test';
    Alert(Alert.defaultProps);
  });

  test('Content', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Content />);
    expect(tree).toMatchSnapshot();
  });
});
