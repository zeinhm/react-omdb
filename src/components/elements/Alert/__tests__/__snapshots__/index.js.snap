// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/elements/Alert Content 1`] = `
<section
  className="alert-component"
/>
`;
